R/data-storage.R

```diff
@@ -0,0 +1,313 @@
+
+#' Data Storage abstract class to handle all the read/write operations
+#'
+#' @description
+#' Abstract R6 Class that encapsulates all the operations needed by
+#' Shiny.telemetry to read and write. This removes the complexity from the
+#' functions and uses a unified API.
+DataStorage <- R6::R6Class( # nolint object_name_linter
+  classname = "DataStorage",
+  #
+  #
+  public = list(
+
+    #' @description initialize data storage object
+    #'
+    #' @param username string with username of the current session
+    #' @param session_id string with custom session id (should not be used)
+
+    initialize = function(username, session_id = NULL) {
+      checkmate::expect_string(username)
+      checkmate::expect_string(session_id, null.ok = TRUE)
+
+      private$.username <- username
+      private$.session_id <- session_id
+
+      # generate if session_id doesn't exist null
+      if (is.null(session_id)) {
+        private$.session_id <- private$generate_session_id()
+      }
+    },
+
+    #' @description Insert new data
+    #'
+    #' @param values list of values to write to storage provider
+    #' @param bucket string with name of type of data to write (example, for
+    #' SQL it should represent a table)
+    #' @param add_username boolean flag that indicates if line should include
+    #' the username of the current session
+
+    insert = function(values, bucket = "user_log", add_username = TRUE) {
+      private$insert_checks(values, bucket, add_username)
+
+      rlang::abort("Method not implemented.")
+    },
+
+    #' @description read all user data
+    #'
+    #' @param date_from date representing the starting day of results
+    #' @param date_to date representing the last day of results
+
+    read_user_data = function(date_from, date_to) {
+      rlang::abort("Method not implemented.")
+    },
+
+    #' @description read all session data
+    #'
+    #' @param date_from date representing the starting day of results
+    #' @param date_to date representing the last day of results
+
+    read_session_data = function(date_from, date_to) {
+      rlang::abort("Method not implemented.")
+    },
+    #' @description read all session data
+    close = function() {
+      private$close_connection()
+    }
+  ),
+  active = list(
+
+    #' @field username string field that returns the current username
+
+    username = function() {
+      private$.username
+    },
+
+    #' @field session_id string field that returns the current session_id
+
+    session_id = function() {
+      private$.session_id
+    }
+  ),
+  private = list(
+    .username = NULL,
+    .session_id = NULL,
+    generate_session_id = function() {
+      uuid::UUIDgenerate()
```

Comment:
I reverted that change and have a branch that could add uuid in the future as an optional dependency

```R
+if (require(uuida, quietly = TRUE)) {
+  return(uuid::UUIDgenerate())
+}

paste(
  c(
    sample(c(letters, LETTERS, 0:9), 10),
    ...
```

I also changed the documentation so that it suggest using Shiny's `session$token` field that is not well documented ([see here](https://github.com/rstudio/shiny/blob/4d276b307e73b1d312833f03a873e5c16a2dad48/R/shiny.R#LL697C5-L697C66))